README.md
# stactools-drcog-lulc

- Name: drcog-lulc
- Package: `stactools.drcog_lulc`
- PyPI: https://pypi.org/project/stactools-drcog-lulc/
- Owner: @pholleway
- Dataset homepage: https://drcog.org/services-and-resources/data-maps-and-modeling/regional-land-use-land-cover-project
- STAC extensions used:
  - [classification](https://github.com/stac-extensions/classification/)
  - [item-assets](https://github.com/stac-extensions/item-assets)
  - [proj](https://github.com/stac-extensions/projection)
  - [raster](https://github.com/stac-extensions/raster)
  - [scientific](https://github.com/stac-extensions/scientific)

This repository will assist you in the generation of STAC files for the Denver Regional Council of Goverments (DRCOG) Land Use Land Cover (LULC) dataset.

DRCOG provides data for 2018 and 2020 in geodatabase form. This package assumes a raster has been extracted from each geodatabase and tiled into smaller cloud optimized GeoTIFF (COG) image files. The tile filenames are expected to contain the prefix `DRCOG_{year}_LULC_`, where `{year}` is either "2018" or "2020".


## Examples

### STAC objects

- [Collection](examples/collection.json)
- [Item](examples/DRCOG_2018_LULC_E3220000_N1710000/DRCOG_2018_LULC_E3220000_N1710000.json)

### Command-line usage

To create a STAC `Item` for a 2018 tile:

```bash
stac drcog-lulc create-item tests/data-files/DRCOG_2018_LULC_E3220000_N1710000.tif examples
```

To create a STAC `Collection` from a file containing a list of DRCOG asset hrefs:

```bash
stac drcog-lulc create-collection examples --file-list examples/file-list.txt
```

The above `create-collection` command is exactly how the contents of the `examples` directory are generated.

Use `stac drcog-lulc --help` to see all subcommands and options.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/). This project attempts to match the major and minor versions of [stactools](https://github.com/stac-utils/stactools) and increments the patch number as needed.

## [0.2.0] - 2022-07-11

### Added

- Item creation for 2018 and 2020 data tiles
- Added a COG tiler ([#6](https://github.com/stactools-packages/drcog-lulc/pull/6))

## [0.1.0] - 2022-05-17

Initial release.

[Unreleased]: <http://github.com/stactools-packages/drcog-lulc/compare/v0.2.0..main>
[0.2.0]: <https://github.com/stactools-packages/drcog-lulc/compare/v0.1.0..v0.2.0>
[0.1.0]: <https://github.com/stactools-packages/drcog-lulc/releases/tag/v0.1.0>


tests/test_stac.py
import unittest

from stactools.drcog_lulc import stac

from . import test_data


class StacTest(unittest.TestCase):
    def test_create_collection(self):
        collection = stac.create_collection()
        collection.set_self_href("")
        self.assertEqual(collection.id, "drcog-lulc")
        collection.validate()

    def test_create_2018(self):
        href = test_data.get_path("data-files/DRCOG_2018_LULC_E3220000_N1710000.tif")
        item = stac.create_item(href)
        self.assertEqual(item.id, "DRCOG_2018_LULC_E3220000_N1710000")

        data_dict = item.assets["data"].to_dict()
        self.assertEqual(data_dict["raster:bands"][0]["unit"], "usft")

        item.validate()

    def test_create_2020(self):
        href = test_data.get_path("data-files/DRCOG_2020_LULC_E2880000_N1680000.tif")
        item = stac.create_item(href)
        self.assertEqual(item.id, "DRCOG_2020_LULC_E2880000_N1680000")

        data_dict = item.assets["data"].to_dict()
        self.assertEqual(data_dict["raster:bands"][0]["unit"], "usft")

        item.validate()


tests/test_module.py
import unittest

import stactools.drcog_lulc


class TestModule(unittest.TestCase):
    def test_version(self):
        self.assertIsNotNone(stactools.drcog_lulc.__version__)


tests/test_commands.py
import os.path
from tempfile import TemporaryDirectory

import pystac
from stactools.testing import CliTestCase

from stactools.drcog_lulc.commands import create_drcog_lulc_command

from . import test_data


class CommandsTest(CliTestCase):
    def create_subcommand_functions(self):
        return [create_drcog_lulc_command]

    def test_create_collection(self):
        with TemporaryDirectory() as temporary_directory:
            result = self.run_command(
                ["drcog-lulc", "create-collection", temporary_directory]
            )
            self.assertEqual(result.exit_code, 0, msg="\n{}".format(result.output))

            jsons = [p for p in os.listdir(temporary_directory) if p.endswith(".json")]
            self.assertEqual(len(jsons), 1)

            collection = pystac.read_file(os.path.join(temporary_directory, jsons[0]))
            self.assertEqual(collection.id, "drcog-lulc")

            collection.validate()

    def test_create_item(self):
        asset_href = test_data.get_path(
            "data-files/DRCOG_2018_LULC_E3220000_N1710000.tif"
        )
        with TemporaryDirectory() as tmp_dir:
            result = self.run_command(
                [
                    "drcog-lulc",
                    "create-item",
                    asset_href,
                    tmp_dir,
                ]
            )
            self.assertEqual(result.exit_code, 0, msg="\n{}".format(result.output))

            jsons = [p for p in os.listdir(tmp_dir) if p.endswith(".json")]
            self.assertEqual(len(jsons), 1)

            item_path = os.path.join(tmp_dir, "DRCOG_2018_LULC_E3220000_N1710000.json")
            item = pystac.read_file(item_path)
            self.assertEqual(item.id, "DRCOG_2018_LULC_E3220000_N1710000")

            item.validate()


src/stactools/drcog_lulc/stac.py
import copy
import logging
import os
from datetime import datetime, timezone
from typing import Optional

import stactools.core.create
from pystac import Asset, Collection, Item
from pystac.extensions.item_assets import ItemAssetsExtension
from pystac.extensions.raster import RasterExtension
from pystac.utils import make_absolute_href
from stactools.core.io import ReadHrefModifier

from stactools.drcog_lulc import constants

logger = logging.getLogger(__name__)


def create_item(
    asset_href: str, read_href_modifier: Optional[ReadHrefModifier] = None
) -> Item:
    """Create a STAC Item with a single Asset for a COG tile of the DRCOG LULC
    classification product.

    Args:
        asset_href (str): An href to a COG containing a tile of classification
            data.
        read_href_modifier (Callable[[str], str]): An optional function to
            modify hrefs (e.g. to add a token to a url).

    Returns:
        Item: STAC Item object representing the landcover tile
    """
    id = os.path.splitext(os.path.basename(asset_href))[0]
    year = int(id.split("_")[1])

    item = stactools.core.create.item(asset_href, read_href_modifier=read_href_modifier)

    item.id = id
    item.common_metadata.start_datetime = datetime(year, 1, 1)
    item.common_metadata.end_datetime = datetime(year, 12, 31, 23, 59, 59)
    item.datetime = None
    item.common_metadata.created = datetime.now(tz=timezone.utc)
    item.common_metadata.description = constants.ITEM_DESCRIPTION[year]
    item.common_metadata.mission = constants.MISSION[year]

    asset_dict = copy.deepcopy(constants.ASSET_PROPS)
    asset_dict["href"] = make_absolute_href(asset_href)
    asset_dict["raster:bands"][0]["spatial_resolution"] = constants.RESOLUTION[year]
    item.add_asset("data", Asset.from_dict(asset_dict))

    RasterExtension.add_to(item)
    item.stac_extensions.append(constants.CLASSIFICATION_SCHEMA)

    item.validate()

    return item


def create_collection(collection_id: str = constants.COLLECTION_ID) -> Collection:
    """Creates a STAC Collection for the 2018 and 2020 DRCOG LULC classification
    product.

    Args:
        collection_id (str): Desired ID for the STAC Collection.

    Returns:
        Collection: The created STAC Collection.
    """
    collection = Collection(
        id=collection_id,
        title=constants.COLLECTION_TITLE,
        description=constants.COLLECTION_DESCRIPTION,
        license=constants.LICENSE,
        keywords=constants.KEYWORDS,
        providers=constants.PROVIDERS,
        extent=constants.EXTENT,
        summaries=constants.SUMMARIES,
    )

    item_assets = ItemAssetsExtension.ext(collection, add_if_missing=True)
    item_assets.item_assets = constants.ITEM_ASSETS

    RasterExtension.add_to(collection)
    collection.stac_extensions.append(constants.CLASSIFICATION_SCHEMA)

    collection.add_links(
        [constants.LICENSE_LINK, constants.REPORT_LINK, constants.DATA_LINK]
    )

    return collection


src/stactools/drcog_lulc/tile.py
import os
import shutil
from glob import glob
from typing import List, Optional

import numpy as np
import rasterio
from stactools.core.utils.subprocess import call


class Tile:
    def __init__(
        self,
        left: float,
        bottom: float,
        right: float,
        top: float,
        nodata: Optional[int] = None,
    ):
        self._left = left
        self._bottom = bottom
        self._right = right
        self._top = top
        self._nodata = nodata

    def subset(self, infile, outdir) -> None:
        base = os.path.splitext(os.path.basename(infile))[0]
        outfile = os.path.join(
            outdir, (f"{base}_E{str(int(self._left))}_N{str(int(self._bottom))}.tif")
        )
        args = [
            "gdal_translate",
            "-of",
            "COG",
            "-co",
            "compress=deflate",
            "-co",
            "blocksize=512",
            "-projwin",
            str(self._left),
            str(self._top),
            str(self._right),
            str(self._bottom),
        ]
        if self._nodata is not None:
            args.extend(["-a_nodata", str(self._nodata)])
        args.append(infile)
        args.append(outfile)
        call(args)


def tile(
    infile: str,
    outdir: str,
    size: int,
    left: int,
    bottom: int,
    nodata: Optional[int] = None,
) -> None:
    """Tiles the given input to a grid."""
    with rasterio.open(infile) as dataset:
        _, _, right, top = dataset.bounds
        tiles = _create_tiles(left, bottom, right, top, size, nodata)
    for tile in tiles:
        tile.subset(infile, outdir)


def _create_tiles(
    left: float,
    bottom: float,
    right: float,
    top: float,
    size: int,
    nodata: Optional[int] = None,
) -> List[Tile]:
    x = left
    y = bottom
    tiles = []
    while x < right:
        while y < top:
            tile = Tile(x, y, min(x + size, right), min(y + size, top), nodata)
            tiles.append(tile)
            y += size
        x += size
        y = bottom
    return tiles


def remove_nodata(indir: str, nodata_dir: str) -> None:
    """Removes TIF files that contain only nodata values to a new directory.

    Args:
        indir (str): Directory containing the TIF files.
        nodata_dir (str): New directory for the TIF files that contain only
            nodata values.
    """
    os.mkdir(nodata_dir)

    tif_files = glob(f"{indir}/*.tif")
    for tif_file in tif_files:
        all_nodata = False
        with rasterio.open(tif_file) as src:
            nodata = src.nodata
            data = src.read(1)
            if np.all((data == nodata)):
                all_nodata = True

        if all_nodata:
            filename = os.path.basename(tif_file)
            shutil.move(tif_file, os.path.join(nodata_dir, filename))
            print("R ", end="", flush=True)
        else:
            print(". ", end="", flush=True)

    print()


src/stactools/drcog_lulc/commands.py
import logging
import os
from typing import Optional

import click
from pystac import CatalogType

from stactools.drcog_lulc import stac
from stactools.drcog_lulc.tile import remove_nodata, tile

logger = logging.getLogger(__name__)


def create_drcog_lulc_command(cli):
    """Creates the stactools-drcog-lulc command line utility."""

    @cli.group(
        "drcog-lulc",
        short_help=("Commands for working with stactools-drcog-lulc"),
    )
    def drcog_lulc():
        pass

    @drcog_lulc.command(
        "create-collection",
        short_help="Creates a STAC collection",
    )
    @click.argument("OUTDIR")
    @click.option("-f", "--file-list", help="File containing list of COG tile HREFs")
    @click.option(
        "--validate/--no-validate",
        help="Validate the collection before saving",
        default=True,
    )
    def create_collection_command(
        outdir: str,
        file_list: Optional[str],
        validate: bool,
    ):
        """Creates a STAC Collection.

        \b
        Args:
            outdir (str): The destination directory.
            file_list (Optional[str]): Text file containing one HREF per line.
                The HREFs should point to COG classification tiles.
            validate (bool): Flag to validate the Collection and all Items.
        """
        collection = stac.create_collection()

        if file_list:
            with open(file_list) as f:
                cog_hrefs = [os.path.abspath(line.strip()) for line in f.readlines()]
            for cog_href in cog_hrefs:
                item = stac.create_item(cog_href)
                collection.add_item(item)

        collection.normalize_hrefs(outdir)
        if validate:
            collection.validate_all()
        collection.make_all_asset_hrefs_relative()
        collection.save(catalog_type=CatalogType.SELF_CONTAINED)

        return None

    @drcog_lulc.command("create-item", short_help="Creates a STAC item")
    @click.argument("INFILE")
    @click.argument("OUTDIR")
    def create_item_command(infile: str, outdir: str):
        """Creates a STAC Item

        \b
        Args:
            infile (str): HREF of a classificaton COG tile.
            outdir (str): Directory that will contain the STAC Item.
        """
        item = stac.create_item(infile)
        item_path = os.path.join(outdir, f"{item.id}.json")
        item.set_self_href(item_path)
        item.make_asset_hrefs_relative()
        item.validate()
        item.save_object(include_self_link=False)

        return None

    @drcog_lulc.command("tile", short_help="Tiles a GeoTIFF to a grid of COGs")
    @click.argument("INFILE")
    @click.argument("OUTDIR")
    @click.argument("ORIGIN_X", type=int)
    @click.argument("ORIGIN_Y", type=int)
    @click.argument("SIZE", type=int)
    @click.option("-n", "--nodata", type=int, help="nodata value")
    def tile_command(
        infile: str,
        outdir: str,
        origin_x: int,
        origin_y: int,
        size: int,
        nodata: Optional[int] = None,
    ) -> None:
        """Tiles an input GeoTIFF to a grid of COGs.

        \b
        Args:
            infile (str): HREF to source GeoTIFF to be tiled
            outdir (str): Directory that will contain the COG tiles
            origin_x (int): X coordinate of the tile grid origin. The origin is
                the lower left corner of the area to be tiled.
            origin_y (int): Y coordinate of the tile grid origin. The origin is
                the lower left corner of the area to be tiled.
            size (int): Tile size in linear units of data, e.g., meters or feet.
            nodata (Optional[int]): nodata value to use for tiled COGs. Only
                necessary if the INFILE GeoTIFF does not contain a correct
                nodata value.
        """
        tile(infile, outdir, size, origin_x, origin_y, nodata)

        return None

    @drcog_lulc.command(
        "remove-nodata-tifs",
        short_help="Removes TIF files that contain only nodata values",
    )
    @click.argument("INDIR")
    @click.option("-n", "--nodata_dir", help="directory to place nodata TIF files")
    def remove_nodata_tifs_command(
        indir: str, nodata_dir: Optional[str] = None
    ) -> None:
        """Moves TIF files that contain only nodata values to a new directory.

        Useful after tiling a large area where many tiles do not intersect valid
        data. The default location for the nodata-only TIF files is a new
        subdirectory named 'nodata_tifs' within the directory of the TIF files
        being examined. Use the --nodata_dir option to override this default
        name and location.

        \b
        Args:
            indir (str): Directory of TIF files to be examined.
            nodata_dir (Optional[str]): Optional directory for the nodata TIF
                files.
        """
        if nodata_dir is None:
            nodata_dir = os.path.join(indir, "nodata_tifs")
        remove_nodata(indir, nodata_dir)

        return None

    return drcog_lulc


src/stactools/drcog_lulc/constants.py
from datetime import datetime
from typing import Any, Dict

from pystac import (
    Extent,
    Link,
    Provider,
    ProviderRole,
    SpatialExtent,
    Summaries,
    TemporalExtent,
)
from pystac.extensions.item_assets import AssetDefinition

ITEM_DESCRIPTION = {
    2018: "2018 Denver Regional Council of Governments (DRCOG) Land Use Land Cover (LULC) at 3.28ft (1m) resolution",  # noqa
    2020: "2020 Denver Regional Council of Governments (DRCOG) Land Use Land Cover (LULC) at 1ft resolution",  # noqa
}
MISSION = {
    2018: "2018 DRCOG LULC pilot study covering 1,000 square miles",
    2020: "2020 DRCOG LULC study covering 6,000 square miles",
}
RESOLUTION = {2018: 1, 2020: 0.3}
CLASSIFICATION_SCHEMA = (
    "https://stac-extensions.github.io/classification/v1.1.0/schema.json"
)

# fmt: off
ASSET_PROPS: Dict[str, Any] = {
    "title": "DRCOG LULC",
    "description": "Denver Regional Council of Governments (DRCOG) Land Use Land Cover (LULC) Classifications",  # noqa
    "type": "image/tiff; application=geotiff; profile=cloud-optimized",
    "roles": ["data"],
    "raster:bands": [
        {
            "description": "Classification values",
            "sampling": "area",
            "data_type": "uint8",
            "unit": "usft",
        }
    ],
    "classification:classes": [
        {"value": 1, "description": "Structures", "color_hint": "FF0000"},
        {"value": 2, "description": "Impervious Surfaces", "color_hint": "B2B2B2"},
        {"value": 3, "description": "Water", "color_hint": "00A9E6"},
        {"value": 4, "description": "Prairie/Grassland/Natural Ground Cover", "color_hint": "C7D79E"},   # noqa
        {"value": 5, "description": "Tree Canopy", "color_hint": "267300"},
        {"value": 6, "description": "Turf/Irrigated Land", "color_hint": "70A800"},
        {"value": 7, "description": "Barren Land", "color_hint": "FFEBAF"},
        {"value": 8, "description": "Cropland", "color_hint": "FFAE42"},
    ],
}
# fmt: on

COLLECTION_ID = "drcog-lulc"
COLLECTION_TITLE = "DRCOG LULC"
COLLECTION_DESCRIPTION = "The Denver Regional Council of Governments (DRCOG) Land Use/Land Cover (LULC) datasets are developed in partnership with Babbit Center for Land and Water Policy and the Chesapeake Conservancy's Conservation Innovation Center (CIC). DRCOG LULC includes 2018 data at 1m (3.28ft) resolution covering 1,000 square miles and 2020 data at 1ft resolution covering 6,000 square miles of the Denver, Colorado region. The data is based on USDA's 1 meter National Agricultural Imagery Program (NAIP) aerial imagery and leaf-off aerial ortho-imagery captured as part of the Denver Regional Aerial Photography Project (6-inch resolution everywhere except the mountainous regions to the west, which came in 1-foot resolution)."  # noqa
LICENSE = "proprietary"
LICENSE_LINK = Link(
    rel="license",
    target="https://drcog.org/legal-terms",
    title="DRCOG Legal Terms",
    media_type="text/html",
)
KEYWORDS = ["Land Cover", "Land Use", "NAIP", "USDA"]
PROVIDERS = [
    Provider(
        name="Denver Regional Council of Governments",
        roles=[
            ProviderRole.PRODUCER,
            ProviderRole.LICENSOR,
        ],
        url=(
            "https://drcog.org/services-and-resources/data-maps-and-modeling/"
            "regional-land-use-land-cover-project"
        ),
    )
]
EXTENT = Extent(
    SpatialExtent(
        [
            [
                -105.93962510864995,
                39.10438697007073,
                -103.66801443832743,
                40.320593119647256,
            ],
            [
                -105.54671456161505,
                39.54013841830152,
                -104.46335720577567,
                39.94430501943824,
            ],
        ]
    ),
    TemporalExtent([[datetime(2018, 1, 1), datetime(2020, 12, 31, 23, 59, 59)]]),
)
SUMMARIES = Summaries({"mission": list(MISSION.values())})
ITEM_ASSETS = {"data": AssetDefinition(ASSET_PROPS)}
REPORT_LINK = Link(
    rel="describedby",
    target="https://gis.drcog.org/rdc/supplemental/lulc_pilot_report.zip",
    title="Supplemental Information",
    media_type="application/zip",
)
DATA_LINK = [
    Link(
        rel="original",
        target="https://landcoverarchive.s3.amazonaws.com/2018/lulc_pilot_raster_2018.zip",
        title="Land_Cover_Raster_Data_2018",
        media_type="application/zip",
    ),
    Link(
        target="https://landcoverarchive.s3.amazonaws.com/2020/DRCOG_2020_Landcover.zip",
        title="Land_Cover_Raster_Data_2020",
        media_type="application/zip",
    ),
]
